### PR TITLE
Convert flow types to TS

### PR DESCRIPTION
**Pull request by saihaj**
### Summary
Convert all the `.js` to `.ts` and fix all syntax related issues that exist because of Flow and TypeScript.

* [x] Fix `%checks` and remove `eslint-disable` comments (https://github.com/saihaj/graphql-js/commit/5d966bd052b692844e729cc83d84ee7ddea06205)
* [x] Change `_` https://github.com/saihaj/graphql-js/commit/fbbb7a557db0f5cd921a0d8b434bdad2c512c40c#r43625283
* [x] use `object` since it is a `WeakMap` (https://github.com/saihaj/graphql-js/commit/ae7c3f406e948ea7411acf02a2f88dc058428213#r43625385)
* [x] Restrict to `integrationTests/*/**` (https://github.com/saihaj/graphql-js/commit/28023e6bf172c63ed5ce00a102fb3c42684cb501)
* [x] Convert  `?` to `Maybe` (few were missed in 41b6975)
* [x] Convert `mixed` to `unknown` (few were missed in cd55ba9)
* [x] Convert to `$Keys` to `keyof`
* [x] Convert `$ReadOnly` to `Readonly`
* [x] Fixup caa8334 into c244b37
* [x] Complete comparison https://github.com/saihaj/graphql-js/pull/1 compare changes made manually with changes made by a tool https://github.com/Khan/flow-to-ts 
* [X] Replace `$ObjMap` to `ObjMap` utility function (need help on this. See discussion here https://github.com/graphql/graphql-js/pull/2828#discussion_r520238458)
* [X] Update `.babelrc-deno` to use TypeScript

Action items Based on https://github.com/graphql/graphql-js/pull/2828#issuecomment-761102728
For @IvanGoncharov 
* [x] Research if we can simplify `visit` typings in `16.x.x`
* [x] Remove polyfills
* [x] Fix test coverage on `16.x.x`
* [ ] https://github.com/graphql/graphql-js/pull/2828#discussion_r554121151
* [ ] Implement `GraphQLAgregatedError`
* [ ] Help @saihaj to generate declarations

For @saihaj 
* [x] Compare and remove .d.ts
* [x] https://github.com/graphql/graphql-js/pull/2828#discussion_r556184614
* [x] Remove Flow infrastructure, even integration test (remove in one commit and split if requested)
* [x] Convert type to interfaces
* [x] Enable as much as possible ESLint rules

~Fix no redeclare `src/polyfills/objectValues.ts`~ (these will be removed so no need)

### CODE REVIEW

**Comment by joostheijkoop** [comment]
Hey! I was asked to check in here. You are doing some serious transformations. Is there any way I can help out? ;)

**Review by 0558436961 — COMMENTED** [review]
يتم اختراق حساب

File: src/jsutils/isCollection.ts [hunk 1]
```diff
@@ -19,10 +19,7 @@
  *   An Object value which might implement the Iterable or Array-like protocols.
  * @return {boolean} true if Iterable or Array-like Object.
  */
-declare function isCollection(value: unknown): boolean;  // FIXME: TS_CONVERTION %check 
-
-// eslint-disable-next-line no-redeclare
-export default function isCollection(obj) {
+export default function isCollection(obj: unknown): obj is object| Symbol {
```
Comment (IvanGoncharov) [inline src/jsutils/isCollection.ts:22]: please squash it to `%checks` commit.

Comment (IvanGoncharov) [reply src/jsutils/isCollection.ts:22]: same for other similar changes.

File: src/language/visitor.ts [hunk 2]
```diff
@@ -17,34 +18,34 @@ export type Visitor<KindToNode, Nodes = $Values<KindToNode>> =
       KindToNode,
       <Node>(Node) => VisitFn<Nodes, Node> | EnterLeave<VisitFn<Nodes, Node>>,
     >;
-type EnterLeave<T> = {| +enter?: T, +leave?: T |};
+type EnterLeave<T> = { readonly enter?: T, readonly leave?: T };
 type ShapeMap<O, F> = $Shape<$ObjMap<O, F>>;
```
Comment (saihaj) [inline src/language/visitor.ts:22]: @IvanGoncharov I am confused on how to convert this`$ObjMap`?

Comment (joostheijkoop) [reply src/language/visitor.ts:22]: Do you mean this? https://github.com/Microsoft/TypeScript/issues/20352

Comment (saihaj) [reply src/language/visitor.ts:22]: Yes but I think there is a utility function for that in here https://github.com/graphql/graphql-js/blob/16.x.x/src/jsutils/ObjMap.js . I did try using that but it didn’t make sense  so I  am little confused

Comment (joostheijkoop) [reply src/language/visitor.ts:22]: It's seems to have one type parameter less, it might work better with de declaration from the linked issue. But need to dig deeper to investigate. Will try tomorrow

Comment (joostheijkoop) [reply src/language/visitor.ts:22]: how do you validate the syntax? I tried runnning `npx typescript` in the checkout of your branch, but to gave a lot (confusing) output

Comment (saihaj) [reply src/language/visitor.ts:22]: yes this is still work in progress and there are many issues that need to be resolved

Comment (IvanGoncharov) [reply src/language/visitor.ts:22]: @saihaj It's two different things:
* `$ObjMap` is Flow utility function https://flow.org/en/docs/types/utilities/#toc-objmap
* `ObjMap` is our utility type: https://github.com/graphql/graphql-js/blob/16.x.x/src/jsutils/ObjMap.js

Comment (IvanGoncharov) [reply src/language/visitor.ts:22]: So  please replace `$ObjMap` with TS equvivalent

**Comment by IvanGoncharov** [comment]
@saihaj Can you please squash these two comments: 
https://github.com/graphql/graphql-js/pull/2828/commits/f47c55f3a878862f4215e35f081abf76ceb271f9
https://github.com/graphql/graphql-js/pull/2828/commits/c5ace4f2a134d1a80b5aaa125ee5979c20472f5c
Just use `git rebase -i` and move later commit earlier and squash them.
Please ping me if you need help with it.

**Comment by saihaj** [comment]
@IvanGoncharov rebased this with `16.x.x.` and squashed those commits

File: src/language/predicates.ts [hunk 3]
```diff
@@ -1,30 +1,41 @@
-import type { ASTNode } from './ast';
+import {
```
Comment (IvanGoncharov) [inline src/language/predicates.ts:1]: should be `import type`

File: src/utilities/getIntrospectionQuery.ts [hunk 4]
```diff
@@ -254,25 +254,25 @@ export type IntrospectionNonNullTypeRef<
 
 export type IntrospectionTypeRef =
   | IntrospectionNamedTypeRef
-  | IntrospectionListTypeRef
+  | IntrospectionListTypeRef<any>
```
Comment (IvanGoncharov) [inline src/utilities/getIntrospectionQuery.ts:257]: TS should support circular refs after 3.7, right?

File: src/type/definition.ts [hunk 5]
```diff
@@ -570,13 +570,12 @@ export class GraphQLScalarType {
...
 
   toConfig(): {
-    ...GraphQLScalarTypeConfig<unknown, unknown>,
     serialize: GraphQLScalarSerializer<unknown>,
     parseValue: GraphQLScalarValueParser<unknown>,
     parseLiteral: GraphQLScalarLiteralParser<unknown>,
     extensions: Maybe<ReadOnlyObjMap<unknown>>,
     extensionASTNodes: ReadonlyArray<ScalarTypeExtensionNode>,
-  } {
+  } & GraphQLScalarTypeConfig<unknown, unknown> {
```
Comment (IvanGoncharov) [inline src/type/definition.ts:583]: change to inheritance after 998bea680d6e11e1c055a400a887a9539de08f75 is added to `16.x.x`

Comment (IvanGoncharov) [reply src/type/definition.ts:583]: you should do that for `definition.ts` and `directives.ts`

File: src/jsutils/PromiseOrValue.ts [hunk 6]
```diff
@@ -1 +1 @@
-export type PromiseOrValue<+T> = Promise<T> | T;
+export type PromiseOrValue<T> = Promise<T> | T;
```
Comment (IvanGoncharov) [inline src/jsutils/PromiseOrValue.ts:1]: Note to myself: I need to check this commit.

File: tsconfig.json [hunk 7]
```diff
@@ -10,5 +10,6 @@
     "types": [],
     "noEmit": true,
     "forceConsistentCasingInFileNames": true
-  }
+  },
+  "exclude": ["integrationTests/*/**"]
 }
```
Comment (IvanGoncharov) [inline tsconfig.json:15]: Superseed by 6e8ca04aff51a47f8378a234f4649f7aeb73514a please remove

File: src/error/GraphQLError.ts [hunk 8]
```diff
@@ -83,7 +83,7 @@ export class GraphQLError extends Error {
     path?: Maybe<ReadonlyArray<string | number>>,
     originalError?: Maybe<(Error & { readonly extensions?: unknown })>,
     extensions?: Maybe<{ [key: string]: unknown }>,
-  ): void {
+  ) {
```
Comment (IvanGoncharov) [inline src/error/GraphQLError.ts:86]: Can you please extract this commit into PR against master?

Comment (saihaj) [reply src/error/GraphQLError.ts:86]: @IvanGoncharov created a new PR https://github.com/graphql/graphql-js/pull/2878. I will remove this commit from here.

File: src/language/visitor.ts (hunk 2 above)
Comment (saihaj) [reply src/language/visitor.ts:22]: > Do you mean this? [microsoft/TypeScript#20352](https://github.com/microsoft/TypeScript/issues/20352)

@IvanGoncharov so as @joostheijkoop mentioned that is the `$ObjMap` is TS equivalent. My questions are:
* Does this utility go in `src/jsutils`
  * Does this need any tests?
* What do I even name this utility since `ObjMap` is already a utility in codebase?
* Or should this change be handled in a new PR once we move `16.x.x` to `main`? all this conversation can be moved to an issue which can be handled by me or someone else in new PR?

File: src/language/visitor.ts [hunk 9]
```diff
@@ -45,7 +45,7 @@ export type VisitFn<TAnyNode, TVisitedNode extends TAnyNode = TAnyNode> = (
  */
 export type VisitorKeyMap<KindToNode> = $ObjMap<
   KindToNode,
-  <T>(T) => ReadonlyArray<$Keys<T>>
+  <T>(T) => ReadonlyArray<keyof T>
```
Comment (IvanGoncharov) [inline src/language/visitor.ts:48]: Should go into next commit, right?

File: src/utilities/findBreakingChanges.ts [hunk 10]
```diff
@@ -65,12 +65,12 @@ export const DangerousChangeType = Object.freeze({
 });
 
 export type BreakingChange = {
-  type: $Keys<typeof BreakingChangeType>;
+  type: keyof typeof BreakingChangeType;
```
Comment (IvanGoncharov) [inline src/utilities/findBreakingChanges.ts:68]: same here

File: src/language/visitor.ts (hunk 9 above)
Comment (IvanGoncharov) [reply src/language/visitor.ts:48]: separate commit

File: src/utilities/getIntrospectionQuery.ts (hunk 4 above)
Comment (saihaj) [reply src/utilities/getIntrospectionQuery.ts:257]: Filed an issue https://github.com/microsoft/TypeScript/issues/42308 and also opened a question on [StackOverflow](https://stackoverflow.com/questions/65693844/using-recursive-type-aliases-gives-circular-reference-error) for help.

File: src/language/blockString.ts [hunk 11]
```diff
@@ -93,8 +93,8 @@ export function getBlockStringIndentation(value: string): number {
  */
 export function printBlockString(
   value: string,
-  indentation?: string = '',
-  preferMultipleLines?: boolean = false,
+  indentation: string = '',
+  preferMultipleLines: boolean = false,
```
Comment (IvanGoncharov) [inline src/language/blockString.ts:97]: Can you please extract it in PR against `master`?

Comment (saihaj) [reply src/language/blockString.ts:97]: @IvanGoncharov moved to https://github.com/graphql/graphql-js/pull/2879

File: src/jsutils/__tests__/didYouMean-test.ts [hunk 12]
```diff
@@ -1,7 +1,7 @@
 import { expect } from 'chai';
 import { describe, it } from 'mocha';
 
-import didYouMean from '../didYouMean';
+import { didYouMean } from '../didYouMean';
```
Comment (IvanGoncharov) [inline src/jsutils/__tests__/didYouMean-test.ts:4]: Can you please remove all default exports in separate PR against `16.x.x`?

Comment (saihaj) [reply src/jsutils/__tests__/didYouMean-test.ts:4]: Opened https://github.com/graphql/graphql-js/pull/2880 will remove this before I rebase

File: src/execution/execute.ts [hunk 13]
```diff
@@ -60,12 +60,13 @@ import {
 import { typeFromAST } from '../utilities/typeFromAST';
 import { getOperationRootType } from '../utilities/getOperationRootType';
 
+import { Maybe } from '../jsutils/Maybe';
+
```
Comment (IvanGoncharov) [inline src/execution/execute.ts:64]: Can you please review commit that introduces `Maybe` and fit it there?
Note: `jsutils` imports should be in the same group and should after `node_modules` imports.

**Comment by IvanGoncharov** [comment]
Some action items for @IvanGoncharov:
* Research if we can simplify `visit` typings in `16.x.x`
* Remove polyfills
* Fix test coverage on `16.x.x`

Action for you @saihaj 
* Compare and remove `.d.ts`
* Remove Flow infrastructure, even integration test (remove in one commit and split if requested)
* Convert type to interfaces
* Enable as much as possible ESLint rules

File: cspell.json [hunk 14]
```diff
@@ -46,6 +47,8 @@
     "Alderaan",
     "Tatooine",
     "astromech",
+    "geocode",
+    "structs",
```
Comment (IvanGoncharov) [inline cspell.json:51]: cspell have some issue with some dictionary marked as applicable on JS but not on TS.
For example:
https://github.com/streetsidesoftware/cspell-dicts/blob/6306b1d11d32e4f66721053d5fccf119ca7ce26e/dictionaries/fullstack/fullstack.txt#L46
https://github.com/streetsidesoftware/cspell-dicts/blob/6306b1d11d32e4f66721053d5fccf119ca7ce26e/dictionaries/fullstack/cspell-ext.json#L23

@saihaj Can you please contribute fixes to cspell?

File: src/type/directives.ts [hunk 15]
```diff
@@ -189,16 +189,16 @@ export const GraphQLDeprecatedDirective = new GraphQLDirective({
 });
 
 /**
- * Used to provide a URL for specifying the behaviour of custom scalar definitions.
+ * Used to provide a URL for specifying the behavior of custom scalar definitions.
```
Comment (IvanGoncharov) [inline src/type/directives.ts:192]: same with `behaviour`

File: src/polyfills/objectEntries.ts [hunk 16]
```diff
@@ -2,7 +2,6 @@ import type { ObjMap } from '../jsutils/ObjMap';
 
 declare function objectEntries<T>(obj: ObjMap<T>): Array<[string, T]>;
 
-/* eslint-disable no-redeclare */
 // $FlowFixMe[name-already-bound] workaround for: https://github.com/facebook/flow/issues/4441
```
Comment (IvanGoncharov) [inline src/polyfills/objectEntries.ts:5]: Can you also remove Flow comments like this one?
Can be in a separate commit.

File: src/error/GraphQLError.ts [hunk 17]
```diff
@@ -5,7 +5,7 @@ import type { Source } from '../language/source';
 import type { SourceLocation } from '../language/location';
 import { getLocation } from '../language/location';
 import { printLocation, printSourceLocation } from '../language/printLocation';
-import { Maybe } from '../jsutils/Maybe';
+import type { Maybe } from '../jsutils/Maybe';
```
Comment (IvanGoncharov) [inline src/error/GraphQLError.ts:8]: Please move it to commit that added `Maybe`

File: src/error/formatError.ts [hunk 18]
```diff
@@ -34,14 +34,14 @@ export type GraphQLFormattedError = {
    * If an error can be associated to a particular point in the requested
    * GraphQL document, it should contain a list of locations.
    */
-  readonly locations: ReadonlyArray<SourceLocation> | void;
+  readonly locations: ReadonlyArray<SourceLocation>;
```
Comment (IvanGoncharov) [inline src/error/formatError.ts:37]: Why did you remove `void`?
This props can have `undefined` as a value

File: src/error/locatedError.ts [hunk 19]
```diff
@@ -1,5 +1,5 @@
 import { inspect } from '../jsutils/inspect';
-import { Maybe } from '../jsutils/Maybe';
+import type { Maybe } from '../jsutils/Maybe';
```
Comment (IvanGoncharov) [inline src/error/locatedError.ts:2]: same

File: src/error/locatedError.ts [hunk 20]
```diff
@@ -12,7 +12,7 @@ import { GraphQLError } from './GraphQLError';
  */
 export function locatedError(
   rawOriginalError: unknown,
-  nodes: ASTNode | ReadonlyArray<ASTNode> | void | null,
+  nodes: ASTNode | ReadonlyArray<ASTNode> | null,
```
Comment (IvanGoncharov) [inline src/error/locatedError.ts:15]: You can pass `undefined` why did you remove it here?

File: src/execution/__tests__/abstract-test.ts [hunk 21]
```diff
@@ -42,6 +42,7 @@ async function executeQuery(args: {
 
 class Dog {
   name: string;
+
```
Comment (IvanGoncharov) [inline src/execution/__tests__/abstract-test.ts:45]: strange, why do you need empty line here?
what `eslint` rule is triggered by it?

File: src/execution/__tests__/abstract-test.ts [hunk 22]
```diff
@@ -52,6 +53,7 @@ class Dog {
 
 class Cat {
   name: string;
+
```
Comment (IvanGoncharov) [inline src/execution/__tests__/abstract-test.ts:56]: same here

File: src/execution/__tests__/lists-test.ts [hunk 23]
```diff
@@ -37,6 +37,7 @@ describe('Execute: Accepts any iterable as list value', () => {
   });
 
   it('Accepts function arguments as a List value', () => {
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (IvanGoncharov) [inline src/execution/__tests__/lists-test.ts:40]: we had setting for it here: [`1ec501c` (#2828)](https://github.com/graphql/graphql-js/pull/2828/commits/1ec501c005fda0ca03614db8ad688a855ebbce0d#diff-9e1ecc14c733bb1ae2e523089f1262ac6ffccbcf950487ee0984403603550e57L630)
You need to remove 1ec501c from PR and just fix eslint errors here.

File: src/execution/execute.ts [hunk 24]
```diff
@@ -262,7 +262,7 @@ export function buildExecutionContext(
   fieldResolver: Maybe<GraphQLFieldResolver<unknown, unknown>>,
   typeResolver?: Maybe<GraphQLTypeResolver<unknown, unknown>>,
 ): ReadonlyArray<GraphQLError> | ExecutionContext {
-  let operation: OperationDefinitionNode | void;
+  let operation: OperationDefinitionNode | undefined;
```
Comment (IvanGoncharov) [inline src/execution/execute.ts:265]: `void` => `undefined` should be a separate commit

**Comment by saihaj** [comment]
I attempted to compare the `.d.ts` files will give one more go just to be sure and then remove them one

File: src/execution/__tests__/abstract-test.ts (hunk 21 above)
Comment (saihaj) [reply src/execution/__tests__/abstract-test.ts]: I might have enabled incorrectly. Now should be updated

**Comment by saihaj** [comment]
Compared and used PR #2899 from @kamilkisiela as a source to fix TS errors. I extracted a few things from that giant commit. That PR removed many `readonly` and there were some discrepancies in types compared to existing definitions which I didn't really understood so I just added `@ts-expect-error: FIXME` which can be fixed in future PRs

File: src/error/__tests__/formatError-test.ts [hunk 25]
```diff
@@ -6,7 +6,6 @@ import { GraphQLError } from '../GraphQLError';
 
 describe('formatError: default error formatter', () => {
   it('uses default message', () => {
-    // $FlowExpectedError[incompatible-call]
```
Comment (IvanGoncharov) [inline src/error/__tests__/formatError-test.ts:9]: @saihaj Please replace instances of `$FlowExpectedError` with `@ts-expect-error`.

File: src/subscription/__tests__/subscribe-test.ts [hunk 26]
```diff
@@ -68,7 +68,7 @@ const EmailEventType = new GraphQLObjectType({
 
 const emailSchema = emailSchemaWithResolvers();
 
-function emailSchemaWithResolvers<T extends unknown>(
+function emailSchemaWithResolvers(
```
Comment (IvanGoncharov) [inline src/subscription/__tests__/subscribe-test.ts:72]: can you please locate the commit where you removed it and fix it there?

File: src/jsutils/__tests__/inspect-test.ts [hunk 27]
```diff
@@ -169,7 +169,7 @@ describe('inspect', () => {
     (Foo.prototype as any)[Symbol.toStringTag] = 'Bar';
     expect(inspect([[new Foo()]])).to.equal('[[[Bar]]]');
 
-    const objectWithoutClassName = new (function () {
+    const objectWithoutClassName = new (function (this) {
```
Comment (IvanGoncharov) [inline src/jsutils/__tests__/inspect-test.ts]: it's incorrect fix, please return eslint disable comment you removed but change them to disable new rule

File: src/__testUtils__/inspectStr.ts [hunk 28]
```diff
@@ -1,4 +1,4 @@
-import { Maybe } from '../jsutils/Maybe';
+import type { Maybe } from '../jsutils/Maybe';
```
Comment (IvanGoncharov) [inline src/__testUtils__/inspectStr.ts:1]: please move this changes into PR where you added `Maybe`

**Comment by IvanGoncharov** [comment]
@saihaj That was a long journey thanks a lot for your help 👍

**Comment by pspeter3** [comment]
Congrats! When will this be released?
